File: tests/upgrade.test.ts
import { describe, expect, test } from 'bun:test'
import {
  buildInstallerEnv,
  type FetchImpl,
  fetchLatestRelease,
  normalizeGitHubReleaseVersion,
  resolveCurrentFactoryInstallDir,
  runLatestInstaller,
  shouldInstallLatest,
  UpgradeError,
} from '../src/upgrade.ts'

function responseFetch(response: Response): FetchImpl {
  return async (_input: string, _init?: RequestInit) => response
}

describe('upgrade version comparison', () => {
  test('normalizes one leading lowercase v', () => {
    expect(normalizeGitHubReleaseVersion('v0.1.0')).toBe('0.1.0')
    expect(normalizeGitHubReleaseVersion('0.1.0')).toBe('0.1.0')
  })

  test('installs only when normalized versions differ', () => {
    expect(shouldInstallLatest('0.1.0', 'v0.1.0')).toBe(false)
    expect(shouldInstallLatest('v0.1.0', '0.1.0')).toBe(false)
    expect(shouldInstallLatest('0.2.0', '0.1.0')).toBe(true)
  })
})

describe('latest release lookup', () => {
  test('parses the latest GitHub release tag', async () => {
    const release = await fetchLatestRelease(responseFetch(Response.json({ tag_name: 'v0.1.0' })))

    expect(release).toEqual({ tagName: 'v0.1.0', version: '0.1.0' })
  })

  test('rejects network failures', async () => {
    const failingFetch: FetchImpl = async () => {
      throw new Error('network down')
    }

    await expect(fetchLatestRelease(failingFetch)).rejects.toThrow(
      'failed to fetch latest GitHub release: network down'
    )
  })

  test('rejects missing or non-string tag_name', async () => {
    await expect(fetchLatestRelease(responseFetch(Response.json({})))).rejects.toThrow(UpgradeError)
    await expect(
      fetchLatestRelease(responseFetch(Response.json({ tag_name: 123 })))
    ).rejects.toThrow(UpgradeError)
  })

  test('rejects invalid JSON as a malformed response', async () => {
    await expect(fetchLatestRelease(responseFetch(new Response('{')))).rejects.toThrow(
      'malformed GitHub release response: invalid JSON'
    )
  })

  test('rejects non-2xx responses', async () => {
    await expect(
      fetchLatestRelease(responseFetch(new Response('nope', { status: 500, statusText: 'bad' })))
    ).rejects.toThrow(UpgradeError)
  })
})

describe('installer execution', () => {
  test('resolves the install directory only for an installed factory binary', () => {
    expect(resolveCurrentFactoryInstallDir('/Users/evan/.local/bin/factory')).toBe(
      '/Users/evan/.local/bin'
    )
    expect(resolveCurrentFactoryInstallDir('/opt/homebrew/bin/bun')).toBeNull()
  })

  test('builds installer env from the parent env', () => {
    const resolved = buildInstallerEnv({ PATH: '/bin', HOME: '/home/evan' }, '/factory/bin')
    expect(resolved['PATH']).toBe('/bin')
    expect(resolved['HOME']).toBe('/home/evan')
    expect(resolved['FACTORY_INSTALL_DIR']).toBe('/factory/bin')

    const unresolved = buildInstallerEnv(
      { PATH: '/bin', FACTORY_INSTALL_DIR: '/old/factory/bin' },
      null
    )
    expect(unresolved['PATH']).toBe('/bin')
    expect(unresolved['FACTORY_INSTALL_DIR']).toBeUndefined()
  })

  test('installer fetch failure throws UpgradeError', async () => {
    const failingFetch: FetchImpl = async () => {
      throw new Error('installer network down')
    }

    await expect(runLatestInstaller({ fetchImpl: failingFetch })).rejects.toThrow(UpgradeError)
  })

  test('installer failure includes captured output', async () => {
    let thrown: unknown
    try {
      await runLatestInstaller({
        fetchImpl: responseFetch(new Response('echo install')),
        runCommand: async () => ({
          stdout: 'installer stdout',
          stderr: 'installer stderr',
          code: 12,
        }),
        parentEnv: { PATH: '/bin' },
        installDir: null,
        cwd: '/tmp',
      })
    } catch (err) {
      thrown = err
    }

    if (!(thrown instanceof UpgradeError)) {
      throw new Error('expected UpgradeError')
    }
    expect(thrown.message).toContain('installer stdout')
    expect(thrown.message).toContain('installer stderr')
  })
})

File: internal/upgrade/upgrade_test.go
package upgrade

import (
	"context"
	"io"
	"net/http"
	"strings"
	"testing"
)

type roundTrip func(*http.Request) (*http.Response, error)

func (r roundTrip) Do(req *http.Request) (*http.Response, error) {
	return r(req)
}

func response(status int, body string) *http.Response {
	return &http.Response{
		StatusCode: status,
		Status:     http.StatusText(status),
		Body:       io.NopCloser(strings.NewReader(body)),
	}
}

func TestVersionComparison(t *testing.T) {
	if NormalizeGitHubReleaseVersion("v0.1.0") != "0.1.0" {
		t.Fatal("expected leading v to be stripped")
	}
	if ShouldInstallLatest("0.1.0", "v0.1.0") {
		t.Fatal("same normalized versions should not upgrade")
	}
	if !ShouldInstallLatest("0.1.0", "0.1.1") {
		t.Fatal("different versions should upgrade")
	}
}

func TestFetchLatestRelease(t *testing.T) {
	release, err := FetchLatestRelease(context.Background(), roundTrip(func(req *http.Request) (*http.Response, error) {
		if req.Header.Get("User-Agent") != "factory" {
			t.Fatalf("missing user agent: %s", req.Header.Get("User-Agent"))
		}
		return response(200, `{"tag_name":"v0.1.0"}`), nil
	}))
	if err != nil {
		t.Fatal(err)
	}
	if release.TagName != "v0.1.0" || release.Version != "0.1.0" {
		t.Fatalf("unexpected release: %+v", release)
	}
}

func TestFetchLatestReleaseRejectsMalformedResponse(t *testing.T) {
	_, err := FetchLatestRelease(context.Background(), roundTrip(func(req *http.Request) (*http.Response, error) {
		return response(200, `{}`), nil
	}))
	if err == nil {
		t.Fatal("expected error")
	}
}

func TestResolveCurrentFactoryInstallDir(t *testing.T) {
	dir := ResolveCurrentFactoryInstallDir("/Users/evan/.local/bin/factory")
	if dir == nil || *dir != "/Users/evan/.local/bin" {
		t.Fatalf("unexpected install dir: %v", dir)
	}
	if ResolveCurrentFactoryInstallDir("/opt/homebrew/bin/bun") != nil {
		t.Fatal("source runner should not resolve install dir")
	}
}

func TestBuildInstallerEnv(t *testing.T) {
	dir := "/factory/bin"
	env := BuildInstallerEnv([]string{"PATH=/bin", "FACTORY_INSTALL_DIR=/old"}, &dir)
	joined := strings.Join(env, "\n")
	if !strings.Contains(joined, "PATH=/bin") || !strings.Contains(joined, "FACTORY_INSTALL_DIR=/factory/bin") {
		t.Fatalf("unexpected env: %#v", env)
	}
	if strings.Contains(joined, "FACTORY_INSTALL_DIR=/old") {
		t.Fatalf("old install dir was not removed: %#v", env)
	}
}
